modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj

```diff
@@ -553,80 +553,64 @@
-                                 ;; timestamp_diff doesn't support months, so convert to datetime to use datetime_diff
-                                 (hsql/call :datetime_diff (hx/->datetime b) (hx/->datetime a) (hsql/raw "month"))
-                                 (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b))))
-                                3))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :month
-      (let [positive-diff (fn [a b] ; precondition: a <= b
-                            (hx/-
-                             ;; timestamp_diff doesn't support months, so convert to datetime to use datetime_diff
-                             (hsql/call :datetime_diff (hx/->datetime b) (hx/->datetime a) (hsql/raw "month"))
-                             (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b)))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :week
-      (let [x (trunc :day x)
-            y (trunc :day y)
-            positive-diff (fn [a b]
-                            (hx/cast
-                             :integer
-                             (hx/floor
-                              (hx// (hsql/call :timestamp_diff b a (hsql/raw "day")) 7))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :day
-      (let [x (trunc :day x)
-            y (trunc :day y)]
-        (hsql/call :timestamp_diff y x (hsql/raw (name unit))))
-
-      (:hour :minute :second)
-      (hsql/call :timestamp_diff y x (hsql/raw (name unit))))))
+  (let [x (sql.qp/->honeysql driver x)
+        y (sql.qp/->honeysql driver y)
+        _ (datetime-diff-check-args x y)]
+    (sql.qp/datetime-diff driver unit x y)))
+
+(defn- timestamp-diff [unit x y]
+  (hsql/call :timestamp_diff y x (hsql/raw (name unit))))
+
+(defmethod sql.qp/datetime-diff [:bigquery-cloud-sdk :year]
+  [driver _unit x y]
+  (hx// (sql.qp/datetime-diff driver :month x y) 12))
+
+(defmethod sql.qp/datetime-diff [:bigquery-cloud-sdk :quarter]
+  [driver _unit x y]
+  (hx// (sql.qp/datetime-diff driver :month x y) 3))
+
+(defmethod sql.qp/datetime-diff [:bigquery-cloud-sdk :month]
+  [_driver _unit x y]
+  ;; Only bigquery's `datetime_diff` supports months. We need to convert args to datetime to use it.
+  (hx/+ (hsql/call :datetime_diff (hx/->datetime y) (hx/->datetime x) (hsql/raw "month"))
+        ;; datetime_diff counts month boundaries not whole months, so we need to adjust
+        ;; if x<y but x>y in the month calendar then subtract one month
+        ;; if x>y but x<y in the month calendar then add one month
+        (hsql/call
+         :case
+         (hsql/call :and (hsql/call :< x y) (hsql/call :> (extract :day x) (extract :day y))) -1
+         (hsql/call :and (hsql/call :> x y) (hsql/call :< (extract :day x) (extract :day y))) 1
+         :else 0)))
```

Comment:
Note this is performing the same calculation logically as before, but optimizing for the length of the SQL expression. The quarter and year calculations are now derived from month calculation and resulting the SQL is the same length or shorter.